firedrake/output.py

```diff
@@ -1,562 +1,502 @@
-            self.addData("UInt8", "types")
-            self.xml.closeElement("PCellData")
-            self.xml.openElement("PPoints")
-            self.addData("Float64", "Points", 3)
-            self.xml.closeElement("PPoints")
-            self._initialised = True
-        vtu_name = os.path.relpath(filepath, start=self.root)
-        self.xml.stream.write('<Piece Source="%s"/>\n' % vtu_name)
-
-    def addData(self, dtype, name, num_of_components=1):
-        """Adds data array description of PDataArray. The header is as follows:
-        <PDataArray type="dtype" Name="name"
-        NumberOfComponents=num_of_components/>"""
-        self.xml.openElement("PDataArray")
-        self.xml.addAttributes(type=dtype, Name=name,
-                               NumberOfComponents=num_of_components)
-        self.xml.closeElement("PDataArray")
-
-
-class _PVDFile(object):
-
-    """Class that represents PVD file."""
-
-    def __init__(self, filename):
-        # Full path to the file without extension.
-        self._filename = filename
-        self._writer = hl.VtkGroup(self._filename)
-        self._warnings = [False, False]
-        # Keep the index of child file
-        # (parallel -> pvtu, else vtu)
-        self._child_index = 0
-        self._time_step = -1
-        # _generate_time -> This file does not accept (function, time) tuple
-        #                   for __lshift__, and it generates the integer
-        #                   time step by itself instead.
-        self._generate_time = False
-
-    def __lshift__(self, data):
-        if isinstance(data, tuple):
-            if self._time_step == -1 or not self._generate_time:
-                self._time_step = data[1]
-                self._update_PVD(data[0])
+            family = "Discontinuous Lagrange"
+
+        output = self._output_functions.get(function)
+        if output is None:
+            # Build appropriate space for output function.
+            shape = function.ufl_shape
+            if len(shape) == 0:
+                V = FunctionSpace(function.ufl_domain(), family, 1)
+            elif len(shape) == 1:
+                if numpy.prod(shape) > 3:
+                    raise ValueError("Can't write vectors with more than 3 components")
+                V = VectorFunctionSpace(function.ufl_domain(), family, 1,
+                                        dim=shape[0])
+            elif len(shape) == 2:
+                if numpy.prod(shape) > 9:
+                    raise ValueError("Can't write tensors with more than 9 components")
+                V = TensorFunctionSpace(function.ufl_domain(), family, 1,
+                                        shape=shape)
             else:
-                raise TypeError(
-                    "You cannot start setting the time by giving a tuple.")
+                raise ValueError("Unsupported shape %s" % (shape, ))
+            output = Function(V)
+            self._output_functions[function] = output
+
```

Comment:
This allocates the output function, which wastes memory if there are many functions (or File objects) that use the same output space V.